packages/serverpod/lib/src/database/database_connection.dart

```diff
@@ -496,6 +508,110 @@ class DatabaseConnection {
+    List<Map<String, Map<String, dynamic>>> previousResultSet,
+  ) async {
+    if (includes == null) return {};
+
+    Map<String, Map<int, List<Map<String, dynamic>>>> resolvedListRelations =
+        {};
+
+    for (var entry in includes.includes.entries) {
+      var include = entry.value;
+      var relationFieldName = entry.key;
+
+      var relativeRelationTable = table.getRelationTable(relationFieldName);
+      var tableRelation = relativeRelationTable?.tableRelation;
+      if (relativeRelationTable == null || tableRelation == null) {
+        throw StateError('Relation table is null this state is impossible.');
+      }
+
+      if (include is IncludeList) {
+        var ids = extractPrimaryKeyForRelation<int>(
+          previousResultSet,
+          tableRelation,
+        );
+
+        if (ids.isEmpty) continue;
+
+        var relationTable = include.table;
+
+        var orderBy = _resolveOrderBy(
+          include.orderByList,
+          include.orderBy,
+          include.orderDescending,
+        );
+
+        var query = SelectQueryBuilder(table: relationTable)
+            .withSelectFields(relationTable.columns)
+            .withWhere(include.where)
+            .withOrderBy(orderBy)
+            .withLimit(include.limit)
+            .withOffset(include.offset)
+            .withWhereRelationInResultSet(ids, relativeRelationTable)
+            .withInclude(include.include)
+            .build();
+
+        print(query);
+
+        var includeListResult = await mappedResultsQuery(session, query);
+
+        var resolvedLists = await _queryIncludedLists(
+          session,
+          include.table,
+          include,
+          includeListResult,
+        );
+
+        var resolvedList = includeListResult
+            .map((rawRow) => resolvePrefixedQueryRow(
+                  relationTable,
+                  rawRow,
+                  resolvedLists,
+                  include: include,
+                ))
+            .whereType<Map<String, dynamic>>()
+            .toList();
+
+        var foreignFieldName = tableRelation.foreignFieldName;
+
+        resolvedListRelations.addAll(_mapListToQueryById(
+          relativeRelationTable,
+          ids,
+          resolvedList,
+          foreignFieldName,
+        ));
+      } else {
+        var resolvedNestedListRelations = await _queryIncludedLists(
+          session,
+          relativeRelationTable,
+          include,
+          previousResultSet,
+        );
+
+        resolvedListRelations.addAll(resolvedNestedListRelations);
+      }
+    }
+
+    return resolvedListRelations;
+  }
+
+  Map<String, Map<int, List<Map<String, dynamic>>>> _mapListToQueryById(
+      Table relativeRelationTable,
+      Set<int> ids,
+      List<Map<String, dynamic>> resolvedList,
+      String foreignFieldName) {
+    return {
+      relativeRelationTable.queryPrefix: {
+        for (var id in ids)
+          id: resolvedList.where((row) => row[foreignFieldName] == id).toList()
```

Comment:
Note: We should probably have this in mind as a performance bottleneck. Much could probably be gained by first sorting the list and the Ids and then mapping them together instead of doing a complete search for every Id.

Reply:
I redid the logic of this!